ocaml/xapi/stream_vdi.ml

```diff
@@ -113,53 +208,95 @@ let send_all refresh_session ofd ~__context rpc session_id (prefix_vdis: vdi lis
+          refresh_session ();
+          let remaining = Int64.sub size offset in
+          if remaining > 0L
+          then
+            begin
+              let this_chunk = (min remaining chunk_size) in
+              let last_chunk = this_chunk = remaining in
+              let this_chunk = Int64.to_int this_chunk in
+              let filename = Printf.sprintf "%s/%08d" prefix chunk_no in
+
+              let now = Unix.gettimeofday () in
+              let time_since_transmission = now -. !last_transmission_time in
+
+              (* We always include the first and last blocks *)
+              let first_or_last = chunk_no = 0 || last_chunk in
+
+              if time_since_transmission > 5. && not first_or_last then begin
+                last_transmission_time := now;
+                write_block ~__context filename Bytes.empty ofd 0;
+                (* no progress has been made *)
+                stream_from (chunk_no + 1) offset
+              end else begin
+                let buffer = if (Int64.of_int this_chunk) = chunk_size
+                  then reusable_buffer
+                  else Bytes.make this_chunk '\000'
+                in
+                Unixext.really_read ifd buffer 0 this_chunk;
+                if not (Zerocheck.is_all_zeros (Bytes.unsafe_to_string buffer) this_chunk) || first_or_last then begin
+                  last_transmission_time := now;
+                  write_block ~__context filename buffer ofd this_chunk;
+                end;
+                made_progress __context progress (Int64.of_int this_chunk);
+                stream_from (chunk_no + 1) (Int64.add offset chunk_size);
+              end
+            end
+        in
+        stream_from 0 0L
+      | Some (path, exportname) ->
+          let actually_write_chunk (this_chunk_no: int) (this_chunk_size: int) =
+            let buffer = if this_chunk_size = Int64.to_int chunk_size
+              then reusable_buffer
+              else Bytes.make this_chunk_size '\000'
+            in
+            let filename = Printf.sprintf "%s/%08d" prefix this_chunk_no in
+            Unix.LargeFile.lseek ifd (Int64.mul (Int64.of_int this_chunk_no) chunk_size) Unix.SEEK_SET |> ignore;
+            Unixext.really_read ifd buffer 0 this_chunk_size;
+            last_transmission_time := Unix.gettimeofday ();
+            write_block ~__context filename buffer ofd this_chunk_size;
+            made_progress __context progress (Int64.of_int this_chunk_size);
+          in
+          let rec stream_from_offset (offset: int64) =
+            let remaining = Int64.sub size offset in
+            if remaining > 0L then begin
+              let this_chunk_size = min (Int64.to_int chunk_size) (Int64.to_int remaining) in
+              let this_chunk_no = Int64.div offset chunk_size in
+              let now = Unix.gettimeofday () in
+              let time_since_transmission = now -. !last_transmission_time in
```

Comment:
I think we prefer to use `Mtime_clock.count` to use a monotonic timer, gettimeofday can go backwards or have large jumps it is a wall clock, we want just relative durations.

Reply:
It would be better to not copy/paste common functionality, but attempt to restructure the code to reuse the existing code instead, I haven't looked to deeply in what the existing code does and how much effort that would be.

Reply:
existing code is looping in a very different way because it can not rely on sparseness. It looks quite hard to refactor in a way that would not be even uglier. ALso as @gaborigloi said, this way we know that we are not introducing bugs to existing code because of this.

Reply:
Good point. The original code had the sime issue

Reply:
I think we could try to fix this mtime issue in our new code because it's a serious issue in my opinion, it could result in intermittent issues.

I agree with Stefano that for the reasons in my above comment, it's best to leave the old code unchanged for now.

Later R3 should fix the old one outside the scope of this req.

Reply:
@edwintorok do you agree with @gaborigloi ? If yes I can do it that way.